content/blog/managing-kubernetes-infrastructure-with-dotnet-and-pulumi/index.md

```diff
@@ -0,0 +1,251 @@
+
+var redisMaster = new ServiceDeployment("redis-master", new ServiceDeploymentArgs
+{
+    Image = "k8s.gcr.io/redis:e2e",
+    Ports = { 6379 },
+});
+
+var redisReplica = new ServiceDeployment("redis-slave", new ServiceDeploymentArgs
+{
+    Image = "gcr.io/google_samples/gb-redisslave:v1",
+    Ports = { 6379 },
+});
+
+var frontend = new ServiceDeployment("frontend", new ServiceDeploymentArgs
+{
+    Replicas = 3,
+    Image = "gcr.io/google-samples/gb-frontend:v4",
+    Ports = { 80 },
+    AllocateIPAddress = true,
+    ServiceType = isMiniKube ? "ClusterIP" : "LoadBalancer",
+});
+
+return new Dictionary<string, object?>{
+    { "frontendIp", frontend.IpAddress },
+};
+```
+
+You can check out the implementation of this `ServiceDeployment` component in the [Guestbook example](https://github.com/pulumi/examples/tree/master/kubernetes-cs-guestbook/components) in the Pulumi Examples repo on GitHub.
+
+## Cloud + Kubernetes with .NET
+
+Because Pulumi lets you work with both Kubernetes and cloud (AWS, Azure, GCP, and more), you can also create and manage the infrastructure that builds both a managed Kubernetes cluster as well as deploying applications and services into the cluster. Using a single familiar programming model, we have access to all these various cloud technologies at our fingertips.
+
+For example, we can deploy a managed Kubernetes cluster on GKE, and then deploy a Pod into it:
+
+```csharp
+var cluster = new KubernetesCluster("do-cluser", new KubernetesClusterArgs {
+    Region = "sfo2",
+    Version = "latest",
+    NodePool = new KubernetesClusterNodePoolArgs {
+        Name = "default",
+        Size = "s-2vcpu-2gb",
+        NodeCount = nodeCount
+    },
+});
+
+var k8sProvider = new Pulumi.Kubernetes.Provider("do-k8s", new Pulumi.Kubernetes.ProviderArgs {
+    KubeConfig = cluster.KubeConfigs.Apply(array => array[0].RawConfig)
+});
+
+var app = new Pulumi.Kubernetes.Apps.V1.Deployment("do-app-dep", new DeploymentArgs {
+    Spec = new DeploymentSpecArgs {
+        Selector = new LabelSelectorArgs {
+            MatchLabels = new InputMap<string>() {
+                {"app", "app-nginx"},
+            },
+        },
+        Replicas = appReplicaCount,
+        Template = new PodTemplateSpecArgs {
+            Metadata = new ObjectMetaArgs() {
+                Labels = new InputMap<string>() {
+                    {"app", "app-nginx"},
+                },
+            },
+            Spec = new PodSpecArgs() {
+                Containers = new ContainerArgs() {
+                    Name = "nginx",
+                    Image = "nginx",
+                }
+            }
+        }
+    },
+}, new CustomResourceOptions {
+    Provider = k8sProvider
+});
+
+var appService = new Pulumi.Kubernetes.Core.V1.Service("do-app-svc", new ServiceArgs {
+    Spec = new ServiceSpecArgs() {
+        Type = "LoadBalancer",
+        Selector = app.Spec.Apply(spec => spec.Template.Metadata.Labels),
```

Comment:
Is the auto-lifting support only for TS? i.e. you can't write this as

`app.Spec.Template.Metadata.Labels`?

Reply:
Lifting is only in JS and Python.  

I think @cyrusnajmabadi and @mikhailshilkov had an idea about how to provide a similar experience in C# (using extension methods), but it requires a much more heavyweight code generation solution. Are we tracking that idea in an issue anywhere?

Reply:
Yes, it's in #3470, which I'm in (early) progress of tearing apart with extended descriptions